src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics.Tracing;
+
+namespace System.Net.Sockets
+{
+    [EventSource(Name = "System.Net.Sockets")]
+    internal sealed class SocketsTelemetry : EventSource
+    {
+        public static readonly SocketsTelemetry Log = new SocketsTelemetry();
+
+        [NonEvent]
+        public static bool IsEnabled(EventLevel level)
+        {
+            return Log.IsEnabled(level, EventKeywords.All);
+        }
+
+        [NonEvent]
+        public void ConnectStart(Internals.SocketAddress address)
+        {
+            ConnectStart(address.ToString());
+        }
+
+        [NonEvent]
+        public void ConnectStart(EndPoint address)
+        {
+            ConnectStart(address.ToString());
+        }
+
+        [Event(1, Level = EventLevel.Informational)]
+        public void ConnectStart(string? address)
+        {
+            WriteEvent(eventId: 1, address ?? "");
+        }
+
+        [Event(2, Level = EventLevel.Informational)]
+        public void ConnectStop()
+        {
+            ConnectStopInternal();
+        }
+
+        [Event(3, Level = EventLevel.Error)]
+        public void ConnectFailed()
+        {
+            WriteEvent(eventId: 3);
+            ConnectStopInternal();
```

Comment:
Maybe part of the strangeness here is: why do we need both events? Why not have a status be part of the stop, rather than logging one event for success and two for exception or cancellation. Also, the exception details don't need to be logged?

Reply:
I assume that having status as part of the stop event can become an issue for Geneva or similar tools queries because it will require event attributes extraction whereas currently they can relate on event ID only. Additionally, this pattern has been already implemented in HttpClient telemetry (RequestStart/Stop/Aborted).

Reply:
@brianrob, can you comment on best practices here?

Reply:
Do you mean that Geneva would need to look at the fields in the event?  I would not expect that to be an issue, and from a perf perspective, it's cheaper to emit the status as a flag in the event than to emit a whole new event.  A reason that one might want to have a separate event for the failure is if the failure occurs at a different call stack than the stop, and that stack is useful for diagnostics.

My recommendation, unless you know that there is an issue with Geneva handling fields (I'm not aware of one), would be that you can put it as a flag on the stop event since the failed and stop events here will have the same stack.  Do make sure to keep the start and stop events so that it's easy to see how long the connection takes, and things work nicely with activity tracking.

Reply:
But what about different EventLevels? Stop/Cancelled/Failed must be logged with different severity. How can it be achieved if we would have the only ConnectStop method with 3 status flags (OK, Cancelled, Failed)?

Reply:
> Stop/Cancelled/Failed must be logged with different severity

Can you elaborate?  Do you mean you expect Canceled/Faulted to be logged even when Stop isn't?

Reply:
Yes, that's correct, Stop is just Informational whereas Cancelled and Failed are of higher severity. It's done the same way as in [HttpTelemetry](https://github.com/dotnet/runtime/blob/87919a0462bc2ebd5c256e5ad7649ff98c858e2c/src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs#L51).

Reply:
IMO it is unlikely we'd get complaints if handling a failed socket connection consumed 1us to send an extra telemetry event. 

Given that start/stop are likely to be much higher volume events than errors I'd imagine customers having the Error level telemetry on all the time whereas start/stop would only get enabled when they needed more detailed performance tracing. That is a good reason to split out error as a separate 3rd event. Correlating the error event back to the stop event does add a little bit of complexity, but it is roughly the same as correlating the stop and start. If our tools have trouble doing this we should probably work to improve our tools rather than duplicate the error data into the stop event.

Reply:
However it is odd to make a single EventSource event API log two events. You might want to define a [NonEvent] API called something like ConnectFailedAndStop() or do two separate calls to the EventSource

Reply:
Could it cause some actual problems? They were combined to avoid redundant code.

Reply:
I'm not aware it will cause any specific problems, just atypical usage that doesn't match suggested patterns others may be familiar with when they try to read your code.

Reply:
Thanks for the idea! I changed it to ConnectFailedAndStop and ConnectCanceledAndStop.